pkg/api/getambassador.io/v2/handwritten.conversion.go

```diff
@@ -9,26 +9,246 @@
 package v2
 
 import (
+	"strings"
+
 	"k8s.io/apimachinery/pkg/conversion"
 
 	"github.com/datawire/ambassador/v2/pkg/api/getambassador.io/v3alpha1"
 )
 
 // These first few functions are written of our own human initiative.
 
+func convert_v2_TLS_To_v3alpha1_TLS(
+	inTLS *BoolOrString,
+	inSvc string,
+
+	outTLS *string,
+	outSvc *string,
+	outExplicit **v3alpha1.V2ExplicitTLS,
+) {
+	// First parse the input.
+	// This parsing logic mimics ircluster.py as closely as possible.
+	originateTLS := false
+	var bareSvc, explicitScheme string
+	switch {
+	case strings.HasPrefix(strings.ToLower(inSvc), "https://"):
+		bareSvc = inSvc[len("https://"):]
+		explicitScheme = inSvc[:len("https://")]
+		originateTLS = true
+	case strings.HasPrefix(strings.ToLower(inSvc), "http://"):
+		bareSvc = inSvc[len("http://"):]
+		explicitScheme = inSvc[:len("http://")]
+	default:
+		bareSvc = inSvc
+	}
+	var ctxName string
+	switch {
+	case inTLS != nil && inTLS.String != nil && *inTLS.String != "":
+		originateTLS = true
+		ctxName = *inTLS.String
+	case inTLS != nil && inTLS.Bool != nil && *inTLS.Bool:
+		originateTLS = *inTLS.Bool
+	}
+	// OK, now re-serialize that in v3alpha1 syntax.
+	if ctxName != "" {
+		*outTLS = *inTLS.String
+		*outSvc = inSvc
+		*outExplicit = &v3alpha1.V2ExplicitTLS{}
+	} else {
+		// 1. outTLS
+		*outTLS = ""
+		var outExplicitTLS string
+		switch {
+		case inTLS == nil:
+			outExplicitTLS = ""
+		case inTLS.Bool != nil && *inTLS.Bool:
+			outExplicitTLS = "bool:true"
+		case inTLS.Bool != nil && !*inTLS.Bool:
+			outExplicitTLS = "bool:false"
+		case inTLS.String != nil && *inTLS.String == "":
+			outExplicitTLS = "string"
+		case *inTLS == BoolOrString{}:
+			outExplicitTLS = "null"
+		}
+
+		// 2. outSvc
+		//
+		//                  | explicitScheme="https://" | explicitScheme="http://" |
+		//  | originateTLS  | inSvc                     | force-https://           |
+		//  | !originateTLS | not possible              | inSvc                    |
+		//
+		// Because an https:// scheme forces originateTLS=true, the bottom-right corner
+		// isn't possible
+		var outExplicitScheme *string
+		if (strings.ToLower(explicitScheme) == "https://") == originateTLS {
```

Comment:
This is really easy to misread, until you look past it and read the "cover the diagonal" comment. Just sayin'. 🙂

Reply:
It's also really easy to mis-think. IDK what to do about that.

Reply:
By the end I was just mashing on the keyboard trying to get the code-coverage and unit tests happy. "What do you mean the `else` case isn't getting covered! The tests are extensive! Ohhh... the `else` case is impossible... how does that make sense?"

Reply:
what about
```go
		schemeIsHTTPS := strings.ToLower(explicitScheme) == "https://"
		if schemeIsHTTPS == originateTLS {
			// cover the diagonal
			*outSvc = inSvc
		} else {
			// cover the top-right
```

Reply:
👍